src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,899 @@
+      // We conclude that ident is not assigned by this target and move on to
+      // the next target.
+    }
+  }
+  return false;
+}
+
+/// Collect identifiers that are local to this loop only
+/// (excluding nested loop).
+static std::unordered_set<irep_idt> gen_loop_locals_set(
+  const dfcc_loop_nesting_grapht &loop_nesting_graph,
+  const std::size_t loop_id)
+{
+  std::unordered_set<irep_idt> loop_locals;
+  for(const auto &target : loop_nesting_graph[loop_id].instructions)
+  {
+    auto loop_id_opt = dfcc_get_loop_id(target);
+    if(
+      target->is_decl() && loop_id_opt.has_value() &&
+      loop_id_opt.value() == loop_id)
+    {
+      loop_locals.insert(target->decl_symbol().get_identifier());
+    }
+  }
+  return loop_locals;
+}
+
+/// Compute subset of locals that must be tracked in the loop's write set.
+/// A local must be tracked if it is dirty or if it may be assigned by one
+/// of the inner loops.
+/// To understand why, just consider what would happen in this situation:
+/// The loop-local must be declared in the assigns clause of the inner loop
+/// otherwise assigns clause checking for the inner loop would fail (from the
+/// point of view of the inner loop, that variable is non-local).
+/// Since write sets inclusion checks are performed between a loop and its
+/// immediately inner loops,  if a loop-local variable
+/// that's dirty or that is assigned by an inner loop is not tracked in the
+/// loop's write set the inclusion check between the loop write set and inner
+/// loop would fail, because the inner loop contains a location that cannot be
+/// found in the outer loop's write set.
+static std::unordered_set<irep_idt> gen_tracked_set(
+  const std::vector<std::size_t> &inner_loops_ids,
+  const std::unordered_set<irep_idt> &locals,
+  dirtyt &dirty,
+  const std::map<std::size_t, dfcc_loop_infot> &loop_info_map)
+{
+  std::unordered_set<irep_idt> tracked;
+  for(const auto &ident : locals)
+  {
+    if(dirty(ident))
+    {
+      tracked.insert(ident);
+    }
+    else
+    {
+      // Check if this ident is touched by one of the inner loops
+      for(const auto inner_loop_id : inner_loops_ids)
+      {
+        if(is_assigned(dirty, ident, loop_info_map.at(inner_loop_id).assigns))
+          tracked.insert(ident);
+      }
+    }
+  }
+  return tracked;
+}
+
+struct contract_clausest
+{
+  exprt invariant_expr;
+  assignst assigns;
+  exprt::operandst decreases_clauses;
+};
+
+/// Generate defaults for all contract clauses of the loop with ID `loop_id` if
+/// at least one type of clause is defined
+static struct contract_clausest default_loop_contract_clauses(
+  const dfcc_loop_nesting_grapht &loop_nesting_graph,
+  const std::size_t loop_id,
+  const irep_idt &function_id,
+  local_may_aliast &local_may_alias,
+  message_handlert &message_handler,
+  const namespacet &ns)
+{
+  messaget log(message_handler);
+
+  const auto loop = loop_nesting_graph[loop_id];
```

Comment:
Nit pick: I believe this will create a copy - don't you want a reference instead?

Reply:
fixed